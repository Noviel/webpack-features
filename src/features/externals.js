import nodeExternals from 'webpack-node-externals';

const reactExternals = [
  {
    react: {
      root: 'React',
      commonjs2: 'react',
      commonjs: 'react',
      amd: 'react',
    },
    'react-dom': {
      root: 'ReactDOM',
      commonjs2: 'react-dom',
      commonjs: 'react-dom',
      amd: 'react-dom',
    },
  },
];

export default (
  { target, production },
  {
    react = false,
    list = [],
    whitelist = [/\.(?!(?:jsx?|json)$).{1,5}$/i],
    modulesDir = 'node_modules',
  }
) => {
  const builtInExternals =
    target.name === 'node'
      ? [
          nodeExternals({
            modulesDir,
            whitelist,
          }),
        ]
      : [];

  return {
    externals: builtInExternals
      .concat(react ? reactExternals : [])
      .concat(list),
  };
};
