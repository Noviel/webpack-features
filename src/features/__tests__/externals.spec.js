import externals from '../externals';
import { callFeature, envs } from '../../fixtures';

describe(`externals`, () => {
  it(`should be an empty array for browsers called without options`, () => {
    const result = callFeature(
      externals,
      undefined,
      [],
      envs.modernBrowsersProd
    );
    expect(result['externals']).toBeInstanceOf(Array);
    expect(result['externals'].length).toBe(0);
  });

  it(`should add react to externals`, () => {
    expect(callFeature(externals, { react: true })).toMatchSnapshot();
  });

  it(`should add default externals for node`, () => {
    const result = callFeature(externals, undefined, [], envs.nodeProd);
    expect(result['externals'][0]).toBeInstanceOf(Function);
    expect(result['externals'].length).toBe(1);
  });
});
