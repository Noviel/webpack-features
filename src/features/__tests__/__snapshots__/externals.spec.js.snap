// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`externals should add react to externals 1`] = `
Object {
  "externals": Array [
    Object {
      "react": Object {
        "amd": "react",
        "commonjs": "react",
        "commonjs2": "react",
        "root": "React",
      },
      "react-dom": Object {
        "amd": "react-dom",
        "commonjs": "react-dom",
        "commonjs2": "react-dom",
        "root": "ReactDOM",
      },
    },
  ],
}
`;

exports[`externals should apply custom 'node_modules' path 1`] = `
Array [
  [Function],
]
`;
